src/ast/codegen_llvm.cpp

```diff
@@ -1499,24 +1516,46 @@ AllocaInst *CodegenLLVM::getMapKey(Map &map)
 {
   AllocaInst *key;
   if (map.vargs) {
-    size_t size = 0;
-    for (Expression *expr : *map.vargs)
+    if (map.vargs->size() == 1)
     {
-      size += expr->type.size;
-    }
-    key = b_.CreateAllocaBPF(size, map.ident + "_key");
-
-    int offset = 0;
-    for (Expression *expr : *map.vargs) {
+      Expression *expr = map.vargs->at(0);
       expr->accept(*this);
-      Value *offset_val = b_.CreateGEP(key, {b_.getInt64(0), b_.getInt64(offset)});
       if (expr->type.type == Type::string || expr->type.type == Type::usym ||
         expr->type.type == Type::inet)
-        b_.CREATE_MEMCPY(offset_val, expr_, expr->type.size, 1);
+      {
+        key = dyn_cast<AllocaInst>(expr_);
+      }
       else
-        // promote map key to 64-bit:
-        b_.CreateStore(b_.CreateIntCast(expr_, b_.getInt64Ty(), expr->type.is_signed), offset_val);
-      offset += expr->type.size;
+      {
+        key = b_.CreateAllocaBPF(expr->type.size, map.ident + "_key");
+        b_.CreateStore(b_.CreateIntCast(expr_, b_.getInt64Ty(), expr->type.is_signed), key);
+      }
+    }
+    else
+    {
+      size_t size = 0;
+      for (Expression *expr : *map.vargs)
+      {
+        size += expr->type.size;
+      }
+      key = b_.CreateAllocaBPF(size, map.ident + "_key");
+
+      int offset = 0;
+      for (Expression *expr : *map.vargs) {
+        expr->accept(*this);
+        Value *offset_val = b_.CreateGEP(key, {b_.getInt64(0), b_.getInt64(offset)});
+        if (expr->type.type == Type::string || expr->type.type == Type::usym ||
```

Comment:
We probably should use `SizedType::IsArray()` here and above to  avoid having to extend even more places when a new array like type gets added

Reply:
I found that using a `Type::array` value as a map key is invalid.

```
./src/bpftrace -e 'BEGIN { @[curtask->numa_faults_locality] = 1; }' 
Attaching 1 probe...
^C

@invalid mapkey argument type
zsh: abort      sudo ./src/bpftrace -e 'BEGIN { @[curtask->numa_faults_locality] = 1; }
```

Should we support `Type::array` as a key and use `SyzeType::IsArray()` here? If not, the semantic analyzer should check the type of keys.

Reply:
Oh in that case leave it as is. Improving Type::array can be done in a different issue